### ISSUE TITLE
Lots of Features

### ISSUE THREAD

**Issue by Scoovy** [nipkownix/re4_tweaks#213]
### Describe the feature you'd like to see added

Me and user SrFuton have been thinking of alot of ideas to add to the Tweaks DLL here are some that we have listed

----Cheat Menu----
God Mode, Infinite HP, Infinite Ammo, Room Swap, Camera Freeze,Rapid Fire, One hit Kill, Infinite Money,Disable Enemy spawn,Krauser Arm everytime,Save/Load Inventory,Inventory Randomizer,Freeze Timer,Character Swap,Disable Laser,In Game Character Coordinates,Player/enemy Size/speed,Press (key) to use Kicks,Auto Aim,Set Character Invincibility,Max HP,Max Upgrade, No Loot Drop,Attach Case Size (S,M,L,XL), Ashley Female Animations (Guns), 

----Other----

Brightness Slider+Gamma Slider (there already is one in the normal re4 but a slider for even higher brightness would be better)
Striker/Matilda Fix (Striker Ditman Fix,Matilda quickturn Fix)
Console On/off (the one you can open normally with F2 on certain Tweaks Versions)
Bring Back Camera Shaking from explosion from gamecube/ PC 07 Version
Customize Color Filters (Video filters in the Video Settings)
Teleportation (aim with laser and press key) (most likely not possible)
Custom Controller Input Settings (press B to run or Press A to Pick items) 

We know some of these are either to hard to add or dont have any reason to be added but seeing them in the DLL would still be nice

**Comment by emoose** [comment]
Did try looking into what Ditman was actually changing a while ago but didn't get that far with it, have documented a lot more of the player/actor/etc related structs since then though so maybe would have better luck now.

My guess is something related to animation speed is getting changed, there's `MOTION_INFO` structs that hold a lot of data about the animation, what might be happening is the `MOTION_INFO` for holding up striker to aim has a faster speed value,  opening up inventory/SubScreen would probably want to change any current `MOTION_INFO` to a different one (the idle inventory anim), but before changing it the game would want to save the current `MOTION_INFO` first & then restore it when exiting inventory, so that the current anim playing in-game wouldn't seem interrupted at all.

The glitch would be that ordinarily the faster-speed anim would have been quickly replaced by a normal-speed one, but the save/restore from inventory ends up restoring the fast-speed anim back onto it, if that makes sense.

That's just a well-informed guess though, not totally sure if the inventory model & in-game model actually share a single `MOTION_INFO` like that, but the games does do a ton of save/restore of different things, so I'd put money on one of them being responsible for it at least.

**Comment by linkthehylian** [comment]
That's super interesting, I had a feeling it was something to do with the animation speed.

So it's basically something like;

- Leon aims Striker before laser is shown while opening inventory (value from aiming Striker is valid)
- Equip another weapon, exit inventory (value is written/saved to Leon upon exiting the inventory)

I still wonder why it's exclusive to the Striker. I'm guessing it might have to do with how fast the aiming animation for that specific weapon is.

**Comment by emoose** [comment]
Oh you have to equip another weapon first? Must have missed that (haven't tried the glitch myself 😄)

If so then the reason might be even easier, maybe aiming the striker has two stages to it, first one sets up something to play a anim at faster speed, and that would normally be immediately followed by another stage that sets it back to a normal anim speed (probably also handles showing laser too)

Opening inventory in between stages & switching weapons cancels it, so only the first stage ends up being ran, and the speed from the fast anim ends up being retained.

This might be backed up by the way weapons are coded too, there's a table of weapon states (and functions for each state), which looks like

```
wepXX_r2_ready
wepXX_r2_set
wepXX_r2_fire
wepXX_r2_down
wepXX_r2_reload
```
I'd guess with ditman only the `ready` func is called for readying the weapon, and `set` ends up being skipped by swapping weapons.

Looking into ready there does seem to be something for changing speed:
![image](https://user-images.githubusercontent.com/1755499/173959845-85ed1d70-0b39-4254-a2ae-2299e93dc497.png)

The value there is at bio4.exe+71CB98 (float) in 1.1.0, maybe changing that could change speed of ditman (seems to only get used by this code & something to do with the difficulty scaling)

**Comment by linkthehylian** [comment]
Yeah, I think that code is exactly what I was looking for then. That's insane how that works lol.

Do you know if any other weapon has that exact code? If not, then it makes sense why it's exclusive to the Striker.

**Comment by emoose** [comment]
Seems pretty unique for the Striker, haven't seen that in any other weapon funcs yet (haven't fully reversed them all tho)

The striker + normal shotgun both actually share the same code, but you can see in that snippet it's checking for striker weapon ID before applying the speed, could be it was just a hack to speed up readying without needing to redo all the anims or something - pretty lucky that they both decided to add it & that someone actually found how to exploit it too.

also just as the code predicted:

https://user-images.githubusercontent.com/1755499/173963874-b00a167b-3468-4be2-8219-828846bee468.mp4

I guess we could probably try looking into into a fix for it (maybe some mod makers wouldn't want people to ditman through their levels :P)

A naive way would be to replace the 1.4 with 1, but that would then slow down the readying anim, and probably break something in the difficulty scaling code.
Maybe a better way would be to just reset the `motInfo.speed` value when changing weapons, kinda surprised they never did that already really (pretty sure there's even a function specifically for resetting anim stuff like that too...)

**Comment by emoose** [comment]
Huh, seems Ditman would also slightly reduce hitbox when playing at 30FPS, sadly it gets broken at 60 though: https://www.youtube.com/watch?v=PcBi8p3z7cw

I'm guessing this is something to do with collision being tied in with velocity, something a little similar to this snippet: https://youtu.be/kpk2tdsPh0A?t=598 (without the out-of-bounds part), but with frame/tick-rate also mixed in somehow, not really sure if we'd be able to help with that at all.
